web/packages/design/src/ButtonWithMenu/ButtonWithMenu.tsx

```diff
@@ -0,0 +1,114 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { ReactElement, useRef, useState } from 'react';
+
+import { ButtonBorder, Flex, Menu, MenuItem } from 'design';
+import * as icons from 'design/Icon';
+import { IconProps } from 'design/Icon/Icon';
+
+/**
+ * Displays a button with a menu to the right of it. Unlike with a regular <Button>, the text of
+ * the button is set through the text prop. <ButtonWithMenu> expects its children to be <MenuItem>
+ * components.
+ *
+ * The icon of the menu can be modified by passing MenuIcon. Props other than text and MenuIcon are
+ * passed to the main <ButtonBorder>. size is automatically passed to both <ButtonBorder> (one for
+ * the main button, one for the menu button) and the icon.
+ *
+ * Clicking on a menu item automatically closes the menu. This is done through event propagation.
+ * Individual menu items can prevent the menu from closing by calling `event.stopPropagation` on the
+ * onClick event.
+ *
+ * @example
+ * <ButtonWithMenu
+ *   text="Text in the button"
+ *   size="large"
+ *   MenuIcon={icons.Cog}
+ * >
+ *   <MenuItem>Foo</MenuItem>
+ *   <MenuItem>Bar</MenuItem>
+ * </ButtonWithMenu>
+ */
+export const ButtonWithMenu = (props: {
+  text: string;
+  children: MenuItemComponent | MenuItemComponent[];
+  MenuIcon?: React.ComponentType<IconProps>;
```

Comment:
I think this `React` type refers to some global variable, let's just import `{ ComponentType }`.